File: rewrite-kotlin/src/test/java/org/openrewrite/kotlin/RemoveImportTest.java
/*
 * Copyright 2023 the original author or authors.
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * https://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.openrewrite.kotlin;

import org.junit.jupiter.api.Test;
import org.openrewrite.DocumentExample;
import org.openrewrite.ExecutionContext;
import org.openrewrite.Recipe;
import org.openrewrite.kotlin.tree.K;
import org.openrewrite.test.RewriteTest;
import org.openrewrite.test.TypeValidation;

import static org.openrewrite.kotlin.Assertions.kotlin;
import static org.openrewrite.test.RewriteTest.toRecipe;

class RemoveImportTest implements RewriteTest {

    static Recipe removeTypeImportRecipe(String type) {
        return toRecipe(() -> new KotlinIsoVisitor<>() {
            @Override
            public K.CompilationUnit visitCompilationUnit(K.CompilationUnit cu, ExecutionContext ctx) {
                maybeRemoveImport(type);
                return cu;
            }
        });
    }

    static Recipe removeMemberImportRecipe(String type, String member) {
        return toRecipe(() -> new KotlinIsoVisitor<>() {
            @Override
            public K.CompilationUnit visitCompilationUnit(K.CompilationUnit cu, ExecutionContext ctx) {
                maybeRemoveImport(type + '.' + member);
                return cu;
            }
        });
    }

    @DocumentExample
    @Test
    void jvmStaticMember() {
        rewriteRun(
          spec -> spec.recipe(removeMemberImportRecipe("java.lang.Integer", "MAX_VALUE")),
          kotlin(
            """
              import java.lang.Integer
              import java.lang.Long

              import java.lang.Integer.MAX_VALUE

              class A
              """,
            """
              import java.lang.Integer
              import java.lang.Long

              class A
              """
          )
        );
    }

    @Test
    void removeStarFoldPackage() {
        rewriteRun(
          spec -> spec.recipe(removeTypeImportRecipe("java.io.OutputStream")).expectedCyclesThatMakeChanges(2),
          kotlin(
            """
              import java.io.*

              class A {
                  val f = File("foo")
              }
              """,
            """
              import java.io.File

              class A {
                  val f = File("foo")
              }
              """
          )
        );
    }

    @Test
    void keepStarFoldPackage() {
        rewriteRun(
          spec -> spec.recipe(removeTypeImportRecipe("java.io.OutputStream")),
          kotlin(
            """
              import java.io.*

              class A {
                  val c : Closeable? = null
                  val f : File? = null
                  val ff : FileFilter? = null
                  val fin : FileInputStream? = null
                  val fos : FileOutputStream? = null
              }
              """
          )
        );
    }

    @Test
    void removeStarFoldTypeMembers() {
        rewriteRun(
          spec -> spec.recipe(removeMemberImportRecipe("java.util.regex.Pattern", "MULTILINE")),
          kotlin(
            """
              import java.util.regex.Pattern.*

              class A {
                  val i = CASE_INSENSITIVE
                  val x = COMMENTS
              }
              """,
            """
              import java.util.regex.Pattern.CASE_INSENSITIVE
              import java.util.regex.Pattern.COMMENTS

              class A {
                  val i = CASE_INSENSITIVE
                  val x = COMMENTS
              }
              """
          )
        );
    }

    @Test
    void keepStarFoldTypeMembers() {
        rewriteRun(
          spec -> spec.recipe(removeMemberImportRecipe("java.util.regex.Pattern", "DOTALL")),
          kotlin(
            """
              import java.util.regex.Pattern.*

              class A {
                  val i = CASE_INSENSITIVE
                  val x = COMMENTS
                  val m = MULTILINE
              }
              """
          )
        );
    }

    @Test
    void keepImportAlias() {
        rewriteRun(
          spec -> spec.typeValidationOptions(TypeValidation.none())
            .recipe(removeMemberImportRecipe("java.util.regex.Pattern", "COMMENTS")),
          kotlin(
            """
              import java.util.regex.Pattern.CASE_INSENSITIVE as i
              import java.util.regex.Pattern.COMMENTS as x

              class A {
                  val f = arrayOf(i, x)
              }
              """
          )
        );
    }

    @Test
    void removeImportAlias() {
        // TODO check if this is really what we want to happen
        rewriteRun(
          // Type validation is disabled until https://github.com/openrewrite/rewrite-kotlin/issues/545 is implemented.
          spec -> spec.typeValidationOptions(TypeValidation.none())
            .recipe(removeMemberImportRecipe("java.util.regex.Pattern", "COMMENTS")),
          kotlin(
            """
              import java.util.regex.Pattern.CASE_INSENSITIVE as i
              import java.util.regex.Pattern.COMMENTS as x

              class A {
                  val f = arrayOf(i)
              }
              """,
            """
              import java.util.regex.Pattern.CASE_INSENSITIVE as i

              class A {
                  val f = arrayOf(i)
              }
              """
          )
        );
    }

    @Test
    void noImportOfImplicitTypes() {
        rewriteRun(
          spec -> spec.recipe(removeTypeImportRecipe("kotlin.Pair")),
          kotlin(
            """
              class A
              """
          )
        );
        rewriteRun(
          spec -> spec.recipe(removeTypeImportRecipe("java.lang.Integer")),
          kotlin(
            """
              class A
              """
          )
        );
    }
}

File: rewrite-core/src/test/kotlin/com/netflix/rewrite/tree/visitor/refactor/op/RemoveImportTest.kt
/**
 * Copyright 2016 Netflix, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.netflix.rewrite.tree.visitor.refactor.op

import com.netflix.rewrite.assertRefactored
import com.netflix.rewrite.parse.OpenJdkParser
import com.netflix.rewrite.parse.Parser
import org.junit.Test

open class RemoveImportTest: Parser by OpenJdkParser() {
    
    @Test
    fun removeNamedImport() {
        val a = parse("""
            import java.util.List;
            class A {}
        """.trimIndent())

        val fixed = a.refactor().removeImport("java.util.List").fix()

        assertRefactored(fixed, "class A {}")
    }

    @Test
    fun removeNamedImportByClass() {
        val a = parse("""
            import java.util.List;
            class A {}
        """.trimIndent())

        val fixed = a.refactor().removeImport("java.util.List").fix()

        assertRefactored(fixed, "class A {}")
    }

    @Test
    fun leaveImportIfRemovedTypeIsStillReferredTo() {
        val a = parse("""
            import java.util.List;
            class A {
               List list;
            }
        """.trimIndent())

        val fixed = a.refactor().removeImport("java.util.List").fix()

        assertRefactored(fixed, """
            import java.util.List;
            class A {
               List list;
            }
        """)
    }

    @Test
    fun removeStarImportIfNoTypesReferredTo() {
        val a = parse("""
            import java.util.*;
            class A {}
        """.trimIndent())

        val fixed = a.refactor().removeImport("java.util.List").fix()

        assertRefactored(fixed, "class A {}")
    }

    @Test
    fun replaceStarImportWithNamedImportIfOnlyOneReferencedTypeRemains() {
        val a = parse("""
            import java.util.*;
            class A {
               Collection c;
            }
        """.trimIndent())

        val fixed = a.refactor().removeImport("java.util.List").fix()

        assertRefactored(fixed, """
            import java.util.Collection;
            class A {
               Collection c;
            }
        """)
    }

    @Test
    fun leaveStarImportInPlaceIfMoreThanTwoTypesStillReferredTo() {
        val a = parse("""
            import java.util.*;
            class A {
               Collection c;
               Set s;
            }
        """.trimIndent())

        val fixed = a.refactor().removeImport("java.util.List").fix()

        assertRefactored(fixed, """
            import java.util.*;
            class A {
               Collection c;
               Set s;
            }
        """)
    }

    @Test
    fun removeStarStaticImport() {
        val a = parse("""
            import static java.util.Collections.*;
            class A {}
        """.trimIndent())

        val fixed = a.refactor().removeImport("java.util.Collections").fix()

        assertRefactored(fixed, "class A {}")
    }

    @Test
    fun leaveStarStaticImportIfReferenceStillExists() {
        val a = parse("""
            import static java.util.Collections.*;
            class A {
               Object o = emptyList();
            }
        """.trimIndent())

        val fixed = a.refactor().removeImport("java.util.Collections").fix()

        assertRefactored(fixed, """
            import static java.util.Collections.*;
            class A {
               Object o = emptyList();
            }
        """)
    }

    @Test
    fun leaveNamedStaticImportIfReferenceStillExists() {
        val a = parse("""
            import static java.util.Collections.emptyList;
            import static java.util.Collections.emptySet;
            class A {
               Object o = emptyList();
            }
        """.trimIndent())

        val fixed = a.refactor().removeImport("java.util.Collections").fix()

        assertRefactored(fixed, """
            import static java.util.Collections.emptyList;
            class A {
               Object o = emptyList();
            }
        """)
    }
}